### PR TITLE
Server code improvements

### PR DESCRIPTION
**Pull request by achim-k**
### Public-Facing Changes

Refactor server code, improve multi-threading safety

### Description
Addresses comments from https://github.com/foxglove/ws-protocol/pull/491. These changes will be applied to that PR once this PR gets merged.

- Fixes modifications to a class member being made while holding only a shared lock
- Log an error if handler functions are not defined by server
- Move text handlers into separate functions
- Other refactoring / simplications

### CODE REVIEW

File: foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp [hunk 1]
```diff
@@ -38,6 +38,29 @@
     }                                                                                          \
   }
 
+namespace {
+
+constexpr uint32_t Integer(const std::string_view str) {
```
Comment (jtbandes) [inline foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp:43]: this function should probably be named StringHash or something that actually explains what it does.

Is there any compile time guarantee that the strings we use it with actually hash to different values? I guess we would hope to get a compile error if it resulted in a switch with multiple identical `case`s?

edit: but I see we are not using switch..

File: foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp [hunk 2]
```diff
@@ -345,7 +372,10 @@ inline void Server<ServerConfiguration>::handleConnectionClosed(ConnHandle hdl)
     _server.get_alog().write(APP, "Client " + clientName + " unadvertising channel " +
                                     std::to_string(clientChannelId) + " due to disconnect");
     if (_handlers.clientUnadvertiseHandler) {
-      _handlers.clientUnadvertiseHandler(clientChannelId, hdl);
+      try {
+        _handlers.clientUnadvertiseHandler(clientChannelId, hdl);
+      } catch (...) {
+      }
```
Comment (jtbandes) [inline foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp:378]: Why do nothing here? Seems like we should log an error if possible?

File: foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp [hunk 3]
```diff
@@ -605,332 +621,46 @@ inline void Server<ServerConfiguration>::handleTextMessage(ConnHandle hdl, Messa
...
-      } catch (const std::exception& e) {
-        sendStatusAndLogMsg(hdl, StatusLevel::Error, e.what());
-      } catch (...) {
-        sendStatusAndLogMsg(hdl, StatusLevel::Error, op + ": Failed to execute handler");
-      }
-    } break;
-    default: {
+  try {
+    const auto intOp = Integer(op);
+    if (intOp == SUBSCRIBE) {
```
Comment (jtbandes) [inline foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp:633]: can this be a switch?

File: foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp [hunk 4]
```diff
@@ -1391,6 +1121,312 @@ inline bool Server<ServerConfiguration>::hasCapability(const std::string& capabi
          _options.capabilities.end();
 }
 
+template <typename ServerConfiguration>
+inline bool Server<ServerConfiguration>::hasHandler(uint32_t op) const {
+  switch (op) {
+    case SUBSCRIBE:
+      return static_cast<bool>(_handlers.subscribeHandler);
```
Comment (jtbandes) [inline foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp:1128]: nit: you can just do `bool(...)`

**Review by jtbandes — APPROVED** [review]
👍

File: foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp (hunk 1 above)
Comment (achim-k) [reply foxglove_bridge_base/include/foxglove_bridge/websocket_server.hpp]: Renamed it and changed it back to a switch

**Pull request merged**